stdlib/LinearAlgebra/src/symmetriceigen.jl

```diff
@@ -164,26 +164,50 @@ function eigen!(A::Hermitian{T,S}, B::Hermitian{T,S}; sortby::Union{Function,Not
     GeneralizedEigen(sorteig!(vals, vecs, sortby)...)
 end
-function eigen!(A::RealHermSymComplexHerm{T,<:StridedMatrix}, B::AbstractMatrix{T}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
-    return _choleigen!(A, B, sortby)
+function eigen(A::RealHermSymComplexHerm{T,<:StridedMatrix}, B::AbstractMatrix{T}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    return eigen!(Matrix{T}(A), eigencopy_oftype(B, T); sortby)
 end
-function eigen!(A::StridedMatrix{T}, B::Union{RealHermSymComplexHerm{T},Diagonal{T}}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
-    return _choleigen!(A, B, sortby)
+function eigen(A::StridedMatrix{T}, B::Union{RealHermSymComplexHerm{T},Diagonal{T}}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    return eigen!(eigencopy_oftype(A, T), Matrix{T}(B); sortby) ;
 end
-function _choleigen!(A, B, sortby)
-    U = cholesky(B).U
-    vals, w = eigen!(UtiAUi!(A, U))
-    vecs = U \ w
+
+function eigen(A::AbstractMatrix{T}, C::Cholesky{T, <:AbstractMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    eigen!(copy(A), C; sortby)
+end
+function eigen!(A::AbstractMatrix{T}, C::Cholesky{T, <:AbstractMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    # Cholesky decomposition based eigenvalues and eigenvectors
+    vals, w = eigen!(UtiAUi!(A, C.U))
+    vecs = C.U \ w
+    GeneralizedEigen(sorteig!(vals, vecs, sortby)...)
+end
+
+function eigen(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    eigen!(copy(A), B; sortby)
+end
+function eigen!(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    # Bunchkaufman decomposition based eigenvalues and eigenvectors
+    if B.uplo == 'U'
+        vals, w = eigen!(ldiv(lu!(copy(B.D)), UiAUti!(A[B.p,B.p], B.U)); sortby)
+        vecs = (B.U' \ w)[invperm(B.p),:]
+    else # B.uplo == 'L'
+        vals, w = eigen!(ldiv(lu!(copy(B.D)), UiAUti!(A[B.p,B.p], B.L)); sortby)
+        vecs = (B.L' \ w)[invperm(B.p),:]
+    end
     GeneralizedEigen(sorteig!(vals, vecs, sortby)...)
 end
 
 # Perform U' \ A / U in-place, where U::Union{UpperTriangular,Diagonal}
-UtiAUi!(A::StridedMatrix, U) = _UtiAUi!(A, U)
+UtiAUi!(A, U) = _UtiAUi!(A, U)
 UtiAUi!(A::Symmetric, U) = Symmetric(_UtiAUi!(copytri!(parent(A), A.uplo), U), sym_uplo(A.uplo))
 UtiAUi!(A::Hermitian, U) = Hermitian(_UtiAUi!(copytri!(parent(A), A.uplo, true), U), sym_uplo(A.uplo))
-
 _UtiAUi!(A, U) = rdiv!(ldiv!(U', A), U)
 
+# Perform U \ A / U' in-place, where U::Union{UpperTriangular,Diagonal}
+UiAUti!(A, U) = _UiAUti!(A, U)
+UiAUti!(A::Symmetric, U) = Symmetric(_UiAUti!(copytri!(parent(A), A.uplo), U), sym_uplo(A.uplo))
+UiAUti!(A::Hermitian, U) = Hermitian(_UiAUti!(copytri!(parent(A), A.uplo, true), U), sym_uplo(A.uplo))
+_UiAUti!(A, U) = rdiv!(ldiv!(U, A), U')
+
```

Comment:
Couldn't we avoid those extra functions by calling `UtiAUi!(A, U')`?

Reply:
@dkarrasch  Thank you for this comment. These functions save a transpose. Would you prefer that I use `UtiAUi!(A,U')` and remove the `UiAUti` functions? Kindly let me know. If so, I will make this change.

Reply:
But the transpose is lazy anyway, so `(U')' === U`, so I don't think there's any performance difference, unless I'm missing something. In that case, I'd say we remove the extra functions.

Reply:
@dkarrasch I had a closer look at this. Based on the following results, I have removed `UiAUti!` and utilised the version with the transpose, as suggested in the comments. Commit: https://github.com/aravindh-krishnamoorthy/julia/commit/0dcd83b00f9853e89f299ae3514b0e2acde77228

```julia
julia> using Revise
julia> using LinearAlgebra
julia> Revise.track(LinearAlgebra)
julia> using BenchmarkTools
julia> A = randn(1000,1000) ;
julia> B = randn(1000,1000) ;
julia> B1 = (B+B')/2 ;
julia> BK = bunchkaufman(B1) ;

julia> methods(LinearAlgebra.UtiAUi!)
# 3 methods for generic function "UtiAUi!" from LinearAlgebra:
 [1] UtiAUi!(A::Hermitian, U)
     @ ~/git/julia/stdlib/LinearAlgebra/src/symmetriceigen.jl:199
 [2] UtiAUi!(A::Symmetric, U)
     @ ~/git/julia/stdlib/LinearAlgebra/src/symmetriceigen.jl:198
 [3] UtiAUi!(A, U)
     @ ~/git/julia/stdlib/LinearAlgebra/src/symmetriceigen.jl:197

julia> methods(LinearAlgebra.UiAUti!)
# 3 methods for generic function "UiAUti!" from LinearAlgebra:
 [1] UiAUti!(A::Hermitian, U)
     @ ~/git/julia/stdlib/LinearAlgebra/src/symmetriceigen.jl:205
 [2] UiAUti!(A::Symmetric, U)
     @ ~/git/julia/stdlib/LinearAlgebra/src/symmetriceigen.jl:204
 [3] UiAUti!(A, U)
     @ ~/git/julia/stdlib/LinearAlgebra/src/symmetriceigen.jl:203

julia> @btime LinearAlgebra._UtiAUi!(A,BK.U) ;
  10.439 ms (4 allocations: 7.64 MiB)

julia> @btime LinearAlgebra._UtiAUi!(A,BK.U') ;
  10.767 ms (5 allocations: 7.64 MiB)

julia> @btime LinearAlgebra._UiAUti!(A,BK.U) ;
  10.770 ms (4 allocations: 7.64 MiB)
```